### ISSUE TITLE
Throw out exception to python when errors happen in c++ code

### ISSUE THREAD

**Issue by shawnLeeZX** [tensorflow/tensorflow#1396]
Required information is after the description.

Hi tensorflow team,

I have enjoyed using tensorflow much. Thanks for the commitment to make this happen.

After writing many code based on it, I found it is very inconvenient that when the C++ code meets an error, it just breaks instead of passing the error to the python front end, so it is impossible to deal with the error gracefully.

For example, when training a network with a large learning rate with ReLU without BN, the activations could easily go out of range. As this is unavoidable since learning rate has to be tuned. It would be great to catch the out of range error in python, so things could be logged and shut down properly, instead of just break.

A log could be

```
W tensorflow/core/common_runtime/executor.cc:1102] 0x33ea350 Compute status: Out of range: Nan in summary histogram for: maxout-relu-cifar10/conv2/HistogramSummary_2
```

Hope this suggestion could be taken into consideration.

### Environment info
Operating System:

```
3.13.0-74-generic #118-Ubuntu SMP x86_64 x86_64 x86_64 GNU/Linux
```

If installed from binary pip package, provide:

1. Which pip package you installed.

```
sudo pip install --upgrade https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow-0.7.1-cp27-none-linux_x86_64.whl
```

2. The output from python -c "import tensorflow; print(tensorflow.__version__)".

```
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcurand.so locally
0.7.1
```

**Comment by vrv** [comment]
Can you explain what currently happens?  Our python code is supposed to return C++ status errors, which can then be handled by users.

**Comment by shawnLeeZX** [comment]
First I have to confess I did not notice the OpError class before. Now I could
get the program log something before it quits. Thanks for pointing it out.

Then I thought a little why I did not notice this, and think this may could be
a small improvement usability. I will elaborate below.

The error with more context is:

```bash
I0309 09:26:19.824918 21189 survivors.py:282] Step 100: loss = nan lr = 0.50000 (359.3 examples/sec 0.356 sec/batch)
W tensorflow/core/common_runtime/executor.cc:1102] 0x2c1c670 Compute status: Out of range: Nan in summary histogram for: maxout-relu-cifar10/conv1/HistogramSummary_2
	 [[Node: maxout-relu-cifar10/conv1/HistogramSummary_2 = HistogramSummary[T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/cpu:0"](maxout-relu-cifar10/conv1/HistogramSummary_2/tag, maxout-relu-cifar10/conv1/biases/read/_470)]]
```

I got the same log whether I catch the error or not. For a typical python
program, when an error is raised, it would print out something like

```
Traceback (most recent call last):
  File "divide_by_zero.py", line 1, in <module>
    1/0
ZeroDivisionError: integer division or modulo by zero
```

If I got something report like it, I would notice I could have error to
catch(it seems a little unreasonble to ask user to read what errors one may get
before coding, since python is a lot about interaction).

So it would be great something could be done on this.

-------------------------------------------------------------------------------

Lastly, I have a side question, in what cases I could get a NaN number? How
could I know?

In this case, this happens when the learning rate is too larger. I guess it is
because the activation value goes out of range(or some bug I do not know in the
convolution op).

I think I could do a denser logging or summary to know what actually happens,
but it would much more convenient when a NaN number is taken as an exception,
since no futher ops can do anything about it(I could file another issue about
this if the team prefers).
